### PR TITLE
add babyai bot and test

### PR DESCRIPTION
**Pull request by GilgameshD**
# Description

Add the babyai bot from the original babyai repo: https://github.com/mila-iqia/babyai/blob/master/babyai/bot.py
This bot can be used to generate expert demonstrations in all babyai environments.

Fixes # (issue)
https://github.com/Farama-Foundation/Minigrid/issues/308

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: minigrid/utils/bot.py [hunk 1]
```diff
@@ -0,0 +1,1028 @@
...
+            self.bot.stack.append(GoNextToSubgoal(self.bot, door_obj, reason="Open"))
+            return
+
+        assert False, "0nothing left to explore"
+
+    def is_exploratory(self):
+        return True
+
+
+class Bot:
```
Comment (pseudo-rnd-thoughts) [inline minigrid/utils/bot.py:559]: As this is for BabyAI only, could we change this name to `BabyAIBot` and change the filename as well to `baby_ai_bot.py`

File: minigrid/utils/bot.py [hunk 2]
```diff
@@ -0,0 +1,1028 @@
...
+    def __str__(self):
+        return repr(self.value)
+
+
+def manhattan_distance(pos, target):
+    return np.abs(target[0] - pos[0]) + np.abs(target[1] - pos[1])
+
+
+class Subgoal:
+    """The base class for all possible Bot subgoals.
```
Comment (pseudo-rnd-thoughts) [inline minigrid/utils/bot.py:36]: The docstring are all of `NumPy` style I believe, pre-commit checks that docstrings are in `google` style. Could you update

File: tests/test_bot.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+def test_bot():
+    # get all babyai envs
+    babyai_envs = []
+    for k_i in gym.envs.registry.keys():
+        if k_i.split("-")[0] == "BabyAI":
+            babyai_envs.append(k_i)
+
+    # random select one env
+    env = gym.make(np.random.choice(babyai_envs), render_mode="human")
```
Comment (pseudo-rnd-thoughts) [inline tests/test_bot.py:17]: Could we `pytest.mark.parameterize` over all BabyAI rather than choose one at random

File: tests/test_bot.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    # reset env
+    env.reset()
+    env.render()
```
Comment (pseudo-rnd-thoughts) [inline tests/test_bot.py:21]: Why do we need to render the environment?

Comment (pseudo-rnd-thoughts) [reply tests/test_bot.py:21]: Also, using `render_mode="human"`

File: tests/test_bot.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+from __future__ import annotations
+
+import gymnasium as gym
+import numpy as np
+
+from minigrid.utils.bot import Bot
+
+
+def test_bot():
```
Comment (pseudo-rnd-thoughts) [inline tests/test_bot.py:9]: What is the goal of the bot? Do we want the bot to solve every environment within the time limit?

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]

**Comment by pseudo-rnd-thoughts** [comment]
Hey, thanks for making the PR, looking through the code, could you look at fixing the suggested points

**Comment by thesofakillers** [comment]
Hey @GilgameshD, are you still working on this?

**Comment by GilgameshD** [comment]
Sorry, I am distracted by other things. I will go back to finish this PR soon.

**Comment by thesofakillers** [comment]
Totally get it! I might try taking over and finishing the PR then if that's alright

**Comment by thesofakillers** [comment]
Can probably close this PR now that #381 is merged :))

**Pull request closed without merging**